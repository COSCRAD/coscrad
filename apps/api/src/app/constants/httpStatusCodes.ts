import { ValueType } from '../../lib/types/valueType';

const httpStatusCodes = {
    ok: 200,
    badRequest: 400,
    unauthorized: 401,
    forbidden: 403,
    notFound: 404,
    internalError: 500,
};

export default httpStatusCodes;

export type HttpStatusCode = ValueType<typeof httpStatusCodes>;
