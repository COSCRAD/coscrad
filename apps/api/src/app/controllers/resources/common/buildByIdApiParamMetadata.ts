export default () => ({
    name: 'id',
    required: true,
    example: '2',
});
