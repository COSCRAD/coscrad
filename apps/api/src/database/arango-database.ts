import { aql, Database } from 'arangojs';
import { Maybe } from '../app/lib/types/maybe';
import { isNotFound, notFound } from '../app/lib/types/not-found';
import { DatabaseProvider } from './database.provider';

type HasKey<T> = T & {
  _key: string;
};

export class ArangoDatabase<TEntity, TCreateEntityDTO, TUpdateEntityDTO> {
  #db: Database;

  constructor(databaseProvider: DatabaseProvider) {
    this.#db = databaseProvider.getConnection();
  }

  fetchById = async (
    id: string,
    collectionName: string
  ): Promise<Maybe<TEntity>> => {
    const collection = this.#getCollection(collectionName);

    if (!collection) return notFound;

    const query = aql`
    FOR t IN ${collection}
      FILTER t._id == @id
        RETURN t
    `;

    const cursor = await this.#db.query(query);

    if (cursor.count === 0) return notFound;

    if (cursor.count > 1)
      throw new Error(
        [
          `More than one document`,
          `with the id ${id}`,
          ` found in collection ${collectionName}`,
        ].join(' ')
      );

    // TODO remove cast
    return cursor.next() as unknown as TEntity;
  };

  fetchMany = async (collectionName: string): Promise<Maybe<TEntity[]>> => {
    const collection = this.#getCollection(collectionName);

    const query = aql`
      FOR t IN ${collection}
        return t
      `;

    const cursor = await this.#db.query(query);

    if (cursor.count === 0) return notFound;

    // TODO remove cast
    return cursor.all() as unknown as TEntity[];
  };

  getCount = async (collectionName: string): Promise<number> => {
    const results = await this.fetchMany(collectionName);

    return isNotFound(results) ? 0 : results.length;
  };

  // Write Methods
  /**
   *
   * TODO [design] consider a pattern for returning errors
   */
  create = async (
    dto: TCreateEntityDTO,
    collectionName: string
  ): Promise<void> => {
    const collection = this.#getCollection(collectionName);

    if (!collection) throw new Error(`Collection ${collectionName} not found!`);

    const query = `
    INSERT @dto
        INTO ${collection}
    `;

    const bindVars = {
      dto,
    };

    await this.#db.query({
      query,
      bindVars,
    });
  };

  createMany = async (
    dtos: TCreateEntityDTO[],
    collectionName: string
  ): Promise<void> => {
    const collection = this.#getCollection(collectionName);

    if (!collection) throw new Error(`Collection ${collectionName} not found!`);

    const query = `
    FOR dto IN @dtos
        INSERT dto
            INTO ${collection}
    `;

    const bindVars = {
      dtos,
    };

    await this.#db.query({
      query,
      bindVars,
    });
  };

  update = async (
    id: string,
    dto: TUpdateEntityDTO,
    collectionName: string
  ): Promise<void> => {
    const collection = this.#getCollection(collectionName);

    if (!collection) throw new Error(`Collection ${collectionName} not found`);

    const documentToUpdate = await this.fetchById(id, collectionName);

    if (isNotFound(documentToUpdate))
      throw new Error(
        [
          `Cannot update document`,
          `${id}`,
          `as no document with that id was found`,
        ].join(' ')
      );

    // TODO remove cast
    const key = this.#getKeyOfDocument(documentToUpdate as HasKey<TEntity>);

    if (isNotFound(key))
      throw new Error(
        `No property '_key' was found on document: ${documentToUpdate}`
      );

    const query = `
    UPDATE ${key}
        WITH @dto
            IN ${collection}
    `;
  };

  // TODO Add Replace

  // TODO Add Soft Delete

  #getCollection(collectionName: string) {
    return this.#db.collection(collectionName);
  }

  #getKeyOfDocument = (document: HasKey<TEntity>): Maybe<string> =>
    typeof document._key === 'string' ? document._key : notFound;
}
