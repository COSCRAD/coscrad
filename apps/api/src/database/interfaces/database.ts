import { EntityId } from '../../domain/types/entity-id';
import { Maybe } from '../../lib/types/maybe';

export interface IDatabase {
  fetchById: <TEntityDTO>(
    id: EntityId,
    collectionName: string
  ) => Promise<Maybe<TEntityDTO>>;

  fetchMany: <TEntityDTO>(
    collectionName: string
  ) => Promise<Maybe<TEntityDTO[]>>;

  create: <TEntityDTO>(
    dto: TEntityDTO,
    collectionName: string
  ) => Promise<void>;

  createMany: <TEntityDTO>(
    dto: TEntityDTO[],
    collectionName: string
  ) => Promise<void>;

  getCount: (collectionName: string) => Promise<number>;

  update: <TEntityDTO>(
    id: EntityId,
    dto: TEntityDTO,
    collectionName: string
  ) => Promise<void>;
}
