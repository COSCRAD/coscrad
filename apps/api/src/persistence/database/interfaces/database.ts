import { ResourceId } from 'apps/api/src/domain/types/ResourceId';
import { Maybe } from 'apps/api/src/lib/types/maybe';

export interface IDatabase {
    fetchById: <TEntityDTO>(id: ResourceId, collectionName: string) => Promise<Maybe<TEntityDTO>>;

    fetchMany: <TEntityDTO>(collectionName: string) => Promise<Maybe<TEntityDTO[]>>;

    create: <TEntityDTO>(dto: TEntityDTO, collectionName: string) => Promise<void>;

    createMany: <TEntityDTO>(dto: TEntityDTO[], collectionName: string) => Promise<void>;

    getCount: (collectionName: string) => Promise<number>;

    update: <TEntityDTO>(id: ResourceId, dto: TEntityDTO, collectionName: string) => Promise<void>;
}
