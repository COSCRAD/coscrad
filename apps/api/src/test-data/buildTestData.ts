import { InMemorySnapshot, resourceTypes } from '../domain/types/resourceTypes';
import buildBookTestData from './buildBookTestData';
import buildEdgeConnectionTestData from './buildEdgeConnectionTestData';
import buildPhotographTestData from './buildPhotographTestData';
import buildSpatialFeatureTestData from './buildSpatialFeatureTestData';
import buildTagTestData from './buildTagTestData';
import buildTermTestData from './buildTermTestData';
import buildTranscribedAudioTestData from './buildTranscribedAudioTestData';
import buildVocabularyListTestData from './buildVocabularyListTestData';

/**
 * **note** When adding new test data \ modifying existing test data, be sure to
 * run `validateTestData.spec.ts` to ensure your test data satisfies all domain
 * invariants.
 */
export default (): InMemorySnapshot => ({
    resources: {
        [resourceTypes.term]: buildTermTestData(),
        [resourceTypes.vocabularyList]: buildVocabularyListTestData(),
        [resourceTypes.transcribedAudio]: buildTranscribedAudioTestData(),
        [resourceTypes.book]: buildBookTestData(),
        [resourceTypes.photograph]: buildPhotographTestData(),
        [resourceTypes.spatialFeature]: buildSpatialFeatureTestData(),
    },
    connections: buildEdgeConnectionTestData(),
    tags: buildTagTestData(),
});
