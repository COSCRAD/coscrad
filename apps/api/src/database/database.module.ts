import { Module } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { DatabaseProvider } from './database.provider';

@Module({
  imports: [DatabaseModule],
  providers: [ConfigService, DatabaseProvider],
  exports: [DatabaseProvider],
})
export class DatabaseModule {}
