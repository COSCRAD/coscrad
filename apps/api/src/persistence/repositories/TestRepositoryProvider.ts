import { Entity } from '../../domain/models/entity';
import { EntityType, InMemorySnapshot } from '../../domain/types/entityType';
import { DatabaseProvider } from '../database/database.provider';
import { getArangoCollectionIDFromEntityType } from '../database/get-arango-collection-ids';
import { RepositoryProvider } from './repository.provider';

export default class TestRepositoryProvider extends RepositoryProvider {
  constructor(databaseProvider: DatabaseProvider) {
    super(databaseProvider);
  }

  // TODO We should correlate entity type with TEntity here
  public async addEntitiesOfSingleType<TEntity extends Entity>(
    entityType: EntityType,
    entities: TEntity[]
  ): Promise<void> {
    await this.forEntity<TEntity>(entityType).createMany(entities);
  }

  // TODO fix types
  public async addEntitiesOfManyTypes(
    snapshot: InMemorySnapshot
  ): Promise<void> {
    const writePromises = Object.entries(snapshot).map(
      ([entityType, entityInstances]) =>
        this.addEntitiesOfSingleType(
          entityType as EntityType,
          entityInstances as any
        )
    );
  }

  public async deleteAllEntitiesOfGivenType(
    entityType: EntityType
  ): Promise<void> {
    await (
      await this.databaseProvider.getDBInstance()
    ).deleteAll(getArangoCollectionIDFromEntityType(entityType));
  }
}
