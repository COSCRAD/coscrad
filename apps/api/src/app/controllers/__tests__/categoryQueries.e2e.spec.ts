import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { ArangoDatabaseProvider } from '../../../persistence/database/database.provider';
import TestRepositoryProvider from '../../../persistence/repositories/__tests__/TestRepositoryProvider';
import generateDatabaseNameForTestSuite from '../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import buildTestData from '../../../test-data/buildTestData';
import httpStatusCodes from '../../constants/httpStatusCodes';
import setUpIntegrationTest from './setUpIntegrationTest';

describe('When fetching the category tree (/treeOfKnowledge)', () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    let app: INestApplication;

    let testRepositoryProvider: TestRepositoryProvider;

    let databaseProvider: ArangoDatabaseProvider;

    const testData = buildTestData().category;

    beforeAll(async () => {
        ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest({
            ARANGO_DB_NAME: testDatabaseName,
        }));
    });

    afterAll(async () => {
        await app.close();

        databaseProvider.close();
    });

    beforeEach(async () => {
        await testRepositoryProvider.testSetup();
    });

    afterEach(async () => {
        await testRepositoryProvider.testTeardown();
    });

    describe(`GET /treeOfKnowledge`, () => {
        it('should return the expected result', async () => {
            await testRepositoryProvider.addCategories(testData);

            const res = await request(app.getHttpServer()).get('/treeOfKnowledge');

            expect(res.status).toBe(httpStatusCodes.ok);

            expect(res.body).toMatchSnapshot();
        });

        /**
         * TODO[https://www.pivotaltracker.com/story/show/184125180]
         *
         * We aren't yet exposing commands through the query endpoint for
         * categories, as its API is different and we haven't created these yet.
         * Once we do so, we should test (as a matter of UX) that we don't
         * return actions to non-admin users, for whom they will certainly fail.
         */
        it.todo('should not return actions for non-admin users');
    });
});
