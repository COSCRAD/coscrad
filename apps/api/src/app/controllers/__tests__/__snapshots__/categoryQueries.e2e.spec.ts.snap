// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When fetching the category tree (/treeOfKnowledge) GET /treeOfKnowledge should return the expected result 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110008",
                  "label": "wolves",
                  "members": Array [
                    Object {
                      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110101",
                      "type": "spatialFeature",
                    },
                  ],
                },
                Object {
                  "children": Array [],
                  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110010",
                  "label": "domestic dogs",
                  "members": Array [
                    Object {
                      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
                      "type": "photograph",
                    },
                  ],
                },
              ],
              "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110004",
              "label": "canines",
              "members": Array [
                Object {
                  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110002",
                  "type": "term",
                },
              ],
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110006",
                  "label": "big cats",
                  "members": Array [],
                },
                Object {
                  "children": Array [],
                  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110007",
                  "label": "domestic cats",
                  "members": Array [
                    Object {
                      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110002",
                      "type": "vocabularyList",
                    },
                    Object {
                      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110000",
                      "type": "photograph",
                    },
                  ],
                },
              ],
              "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110005",
              "label": "felines",
              "members": Array [],
            },
            Object {
              "children": Array [],
              "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110011",
              "label": "rodents",
              "members": Array [
                Object {
                  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110024",
                  "type": "book",
                },
              ],
            },
          ],
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110002",
          "label": "mammals",
          "members": Array [
            Object {
              "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
              "type": "term",
            },
            Object {
              "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110023",
              "type": "book",
            },
            Object {
              "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
              "type": "song",
            },
          ],
        },
        Object {
          "children": Array [],
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110003",
          "label": "birds",
          "members": Array [],
        },
      ],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
      "label": "animals",
      "members": Array [
        Object {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
          "type": "vocabularyList",
        },
      ],
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110015",
                  "label": "tools",
                  "members": Array [
                    Object {
                      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110110",
                      "type": "audioItem",
                    },
                  ],
                },
              ],
              "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110016",
              "label": "adornments",
              "members": Array [],
            },
            Object {
              "children": Array [],
              "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110017",
              "label": "clothing",
              "members": Array [
                Object {
                  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
                  "type": "note",
                },
              ],
            },
            Object {
              "children": Array [],
              "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110018",
              "label": "tools",
              "members": Array [],
            },
          ],
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110013",
          "label": "props",
          "members": Array [],
        },
        Object {
          "children": Array [],
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110014",
          "label": "wardrobe",
          "members": Array [],
        },
      ],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110012",
      "label": "film",
      "members": Array [],
    },
  ],
  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110000",
  "label": "tree of knowledge",
  "members": Array [],
}
`;
